proptest-derive/tests/params.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2018 The proptest developers
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(attr_literals)]
+
+#[macro_use]
+extern crate proptest_derive;
+#[macro_use]
+extern crate proptest;
+use proptest::prelude::{any_with, Arbitrary};
+
+struct ComplexType {
+    max: u64,
+}
+
+impl Default for ComplexType {
+    fn default() -> Self { Self { max: 10 } }
+}
+
+#[derive(Debug, Arbitrary)]
+#[proptest(params(ComplexType))]
+struct TopHasParams {
+    string: usize,
+    #[proptest(strategy = "0..params.max")]
+    int: u64,
+}
+
+#[derive(Debug, Arbitrary)]
+#[proptest(no_params)]
+struct TopNoParams {
+    stuff: usize,
+}
+
+#[derive(Debug, Arbitrary)]
+struct InnerNoParams {
+    string: String,
+    #[proptest(no_params)]
+    has: TopHasParams,
+}
+
+#[derive(Debug, Arbitrary)]
+#[proptest(params(u64))]
+struct TPIS {
+    #[proptest(strategy = "\"a+\"")]
```

Comment:
How hard would it be to have a `#[proptest(regex = "string")]` for `String` and `Vec<[u8]>` fields? It would make syntax like this nicer

Reply:
Not very hard I think; you could do this by adding another attribute `regex` to the parser and then desugaring that to the same internals as `strategy` uses in the attribute parser.

I can do this in a follow up PR if you like :)

(I agree `#[proptest(strategy = "\"a+\"")]` looks suboptimal)